### PR TITLE
[SPARK-49661][SQL] Implement trim collation hashing and comparison.

### PR DESCRIPTION
**Pull request by jovanpavl-db**
### What changes were proposed in this pull request?
Implement support for hashing and comparison for trim collation.

### Why are the changes needed?
To have full support for trim collation.

### How was this patch tested?
Add tests in CollationFactorySUite and CollationSqlExpressionSuite.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 1]
```diff
@@ -154,16 +154,24 @@ public static class Collation {
...
         Collator collator,
         Comparator<UTF8String> comparator,
         String version,
         ToLongFunction<UTF8String> hashFunction,
+        BiFunction<UTF8String, UTF8String, Boolean> equalsFunction,
         boolean supportsBinaryEquality,
         boolean supportsBinaryOrdering,
-        boolean supportsLowercaseEquality) {
+        boolean supportsLowercaseEquality,
+        boolean supportsSpaceTrimming) {
```
Comment (mihailom-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:174]: Is it possible to have binaryEquality and spaceTrimming at the same time? If not could we add another assert?

**Review by mihailomilosevic2001 — COMMENTED** [review]
Seems good to me, but @uros-db would need to take a look as well, as he has more context on hash aggregates and hash with collations in general.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 1 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:174]: this is a good point, for example if we have "UTF8_BINARY_RTRIM", then it's no longer "binaryEquality" per se

however, I suggest we visit this issue in a follow-up, additional assert could be good

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 2]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
+          Comparator<UTF8String> comparator;
+          ToLongFunction<UTF8String> hashFunction;
+          BiFunction<UTF8String, UTF8String, Boolean> equalsFunction;
+          boolean supportsSpaceTrimming = spaceTrimming != SpaceTrimming.NONE;
+
+          if(spaceTrimming == SpaceTrimming.NONE) {
+            comparator = UTF8String::binaryCompare;
+            hashFunction = s -> (long) s.hashCode();
+            equalsFunction = UTF8String::equals;
+          }else {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:554]:
```suggestion
          } else {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 3]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
+            comparator = (s1, s2) -> applyTrimmingPolicy(s1, spaceTrimming).binaryCompare(
+                    applyTrimmingPolicy(s2, spaceTrimming));
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:556]:
```suggestion
            comparator = (s1, s2) -> applyTrimmingPolicy(s1, spaceTrimming).binaryCompare(
              applyTrimmingPolicy(s2, spaceTrimming));
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 4]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
+            hashFunction = s -> (long) applyTrimmingPolicy(s, spaceTrimming).hashCode();
+            equalsFunction = (s1, s2) -> applyTrimmingPolicy(s1, spaceTrimming).equals(
+                    applyTrimmingPolicy(s2, spaceTrimming));
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:559]:
```suggestion
            equalsFunction = (s1, s2) -> applyTrimmingPolicy(s1, spaceTrimming).equals(
              applyTrimmingPolicy(s2, spaceTrimming));
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 5]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
+            /* supportsLowercaseEquality = */ false,
+            spaceTrimming != SpaceTrimming.NONE);
         } else {
+          Comparator<UTF8String> comparator;
+          ToLongFunction<UTF8String> hashFunction;
+
+          if(spaceTrimming == SpaceTrimming.NONE ) {
+            comparator = CollationAwareUTF8String::compareLowerCase;
+            hashFunction = s ->
+                    (long) CollationAwareUTF8String.lowerCaseCodePoints(s).hashCode();
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:581]:
```suggestion
            hashFunction = s ->
              (long) CollationAwareUTF8String.lowerCaseCodePoints(s).hashCode();
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 6]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
             /* supportsBinaryEquality = */ true,
             /* supportsBinaryOrdering = */ true,
-            /* supportsLowercaseEquality = */ false);
+            /* supportsLowercaseEquality = */ false,
+            spaceTrimming != SpaceTrimming.NONE);
         } else {
+          Comparator<UTF8String> comparator;
+          ToLongFunction<UTF8String> hashFunction;
+
+          if(spaceTrimming == SpaceTrimming.NONE ) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:578]:
```suggestion
          if (spaceTrimming == SpaceTrimming.NONE ) {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 7]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
+            comparator = CollationAwareUTF8String::compareLowerCase;
+            hashFunction = s ->
+                    (long) CollationAwareUTF8String.lowerCaseCodePoints(s).hashCode();
+          }else{
+            comparator = (s1, s2) -> CollationAwareUTF8String.compareLowerCase(
+                    applyTrimmingPolicy(s1, spaceTrimming),
+                    applyTrimmingPolicy(s2, spaceTrimming));
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:585]:
```suggestion
            comparator = (s1, s2) -> CollationAwareUTF8String.compareLowerCase(
              applyTrimmingPolicy(s1, spaceTrimming),
              applyTrimmingPolicy(s2, spaceTrimming));
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 8]
```diff
@@ -917,16 +957,35 @@ protected Collation buildCollation() {
         Collator collator = Collator.getInstance(resultLocale);
         // Freeze ICU collator to ensure thread safety.
         collator.freeze();
+
+        Comparator<UTF8String> comparator;
+        ToLongFunction<UTF8String> hashFunction;
+
+        if(spaceTrimming == SpaceTrimming.NONE){
+          hashFunction = s -> (long) collator.getCollationKey(
+                  s.toValidString()).hashCode();
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:966]:
```suggestion
          hashFunction = s -> (long) collator.getCollationKey(
            s.toValidString()).hashCode();
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 9]
```diff
@@ -917,16 +957,35 @@ protected Collation buildCollation() {
...
+          comparator = (s1, s2) ->
+                  collator.compare(s1.toValidString(), s2.toValidString());
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:968]:
```suggestion
          comparator = (s1, s2) ->
            collator.compare(s1.toValidString(), s2.toValidString());
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 10]
```diff
@@ -917,16 +957,35 @@ protected Collation buildCollation() {
...
+        } else {
+          comparator = (s1, s2) -> collator.compare(
+                  applyTrimmingPolicy(s1, spaceTrimming).toValidString(),
+                  applyTrimmingPolicy(s2, spaceTrimming).toValidString());
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:972]:
```suggestion
          comparator = (s1, s2) -> collator.compare(
            applyTrimmingPolicy(s1, spaceTrimming).toValidString(),
            applyTrimmingPolicy(s2, spaceTrimming).toValidString());
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 11]
```diff
@@ -917,16 +957,35 @@ protected Collation buildCollation() {
...
+          hashFunction = s -> (long) collator.getCollationKey(
+                  applyTrimmingPolicy(s, spaceTrimming).toValidString()).hashCode();
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:974]:
```suggestion
          hashFunction = s -> (long) collator.getCollationKey(
            applyTrimmingPolicy(s, spaceTrimming).toValidString()).hashCode();
```

**Review by uros-db — CHANGES_REQUESTED** [review]
left some style-related comments, please fix this up for the 2nd round of review

also, let's please pause this until Monday

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 1 above)
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:174]: Yes, actually i thought about this and plan to address it in follow-up pr.
So yes, this name is not accurate any more. One could say, why just not then set it on false for UTF8_BINARY collations using trimming? Well for many expressions trim specifier is passthrough and we would like to have it done together with UTF8_BINARY collation. Not to mention on many places supportsBinaryEquality = isUtfBinary which is ok i guess because I guess it was not planeed to extend UTF8_binary collaiton.
So simple solution for this would be to **rename supportsBinaryOrdering to something like "isUtf8BinaryType"**.

Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:174]: I created Jira task for this:
https://issues.apache.org/jira/browse/SPARK-49911

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 12]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
       @Override
       protected Collation buildCollation() {
         if (caseSensitivity == CaseSensitivity.UNSPECIFIED) {
+          Comparator<UTF8String> comparator;
+          ToLongFunction<UTF8String> hashFunction;
+          BiFunction<UTF8String, UTF8String, Boolean> equalsFunction;
+          boolean supportsSpaceTrimming = spaceTrimming != SpaceTrimming.NONE;
+
+          if(spaceTrimming == SpaceTrimming.NONE) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:550]:
```suggestion
          if (spaceTrimming == SpaceTrimming.NONE) {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 13]
```diff
@@ -538,27 +542,63 @@ private static boolean isValidCollationId(int collationId) {
...
             /* supportsBinaryEquality = */ true,
             /* supportsBinaryOrdering = */ true,
-            /* supportsLowercaseEquality = */ false);
+            /* supportsLowercaseEquality = */ false,
+            spaceTrimming != SpaceTrimming.NONE);
         } else {
+          Comparator<UTF8String> comparator;
+          ToLongFunction<UTF8String> hashFunction;
+
+          if (spaceTrimming == SpaceTrimming.NONE ) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:578]:
```suggestion
          if (spaceTrimming == SpaceTrimming.NONE) {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 14]
```diff
@@ -1137,7 +1188,7 @@ public static String[] getICULocaleNames() {
 
   public static UTF8String getCollationKey(UTF8String input, int collationId) {
     Collation collation = fetchCollation(collationId);
-    if (usesTrimCollation(collationId)) {
+    if (collation.supportsSpaceTrimming) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:1191]: so we did end up going with this :)

Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:1191]: Yep :)

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala [hunk 15]
```diff
@@ -127,22 +127,26 @@ class CollationFactorySuite extends AnyFunSuite with Matchers { // scalastyle:ig
...
       CollationTestCase("UTF8_LCASE", "aaa", "AaA", true),
       CollationTestCase("UTF8_LCASE", "aaa", "aa", false),
       CollationTestCase("UTF8_LCASE", "aaa", "bbb", false),
       CollationTestCase("UTF8_LCASE", "å", "a\u030A", false),
+      CollationTestCase("UTF8_LCASE_RTRIM", "aaa  ", "AaA", true),
       CollationTestCase("UNICODE", "aaa", "aaa", true),
       CollationTestCase("UNICODE", "aaa", "AAA", false),
       CollationTestCase("UNICODE", "aaa", "bbb", false),
       CollationTestCase("UNICODE", "å", "a\u030A", true),
+      CollationTestCase("UNICODE_RTRIM", "aaa", "aaa  ", true),
```
Comment (uros-db) [inline common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:143]: here too

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala [hunk 16]
```diff
@@ -162,19 +166,32 @@ class CollationFactorySuite extends AnyFunSuite with Matchers { // scalastyle:ig
...
+      CollationTestCase("UNICODE_RTRIM", "aaa ", "aaa  ", 0),
+      CollationTestCase("UNICODE_RTRIM", "aaa", "AAA ", -1),
+      CollationTestCase("UNICODE_RTRIM", "aaa ", "bbb  ", -1),
       CollationTestCase("UNICODE_CI", "aaa", "aaa", 0),
       CollationTestCase("UNICODE_CI", "aaa", "AAA", 0),
-      CollationTestCase("UNICODE_CI", "aaa", "bbb", -1))
+      CollationTestCase("UNICODE_CI", "aaa", "bbb", -1),
+      CollationTestCase("UNICODE_CI_RTRIM", "aaa", "aaa  ", 0),
+      CollationTestCase("UNICODE_CI_RTRIM", "aaa ", "AAA", 0),
+      CollationTestCase("UNICODE_CI_RTRIM", "aaa ", "bbb  ", -1)
```
Comment (uros-db) [inline common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:193]: and of course same goes here

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 17]
```diff
@@ -41,14 +41,11 @@ class StringType private (val collationId: Int) extends AtomicType with Serializ
   private[sql] def supportsBinaryEquality: Boolean =
     CollationFactory.fetchCollation(collationId).supportsBinaryEquality
 
-  private[sql] def supportsLowercaseEquality: Boolean =
-    CollationFactory.fetchCollation(collationId).supportsLowercaseEquality
```
Comment (uros-db) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:45]: why are we removing this?

Comment (jovanpavl-db) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:45]: Not used anywhere.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 18]
```diff
@@ -415,7 +415,7 @@ abstract class HashExpression[E] extends Expression {
 
   protected def genHashString(
       ctx: CodegenContext, stringType: StringType, input: String, result: String): String = {
-    if (stringType.supportsBinaryEquality) {
+    if (stringType.supportsBinaryEquality && !stringType.usesTrimCollation) {
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: I'm just curious - do we have tests that fail if you remove this change?

Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: if not, let's make sure to add them

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 19]
```diff
@@ -566,7 +566,7 @@ abstract class InterpretedHashFunction {
         hashUnsafeBytes(a, Platform.BYTE_ARRAY_OFFSET, a.length, seed)
       case s: UTF8String =>
         val st = dataType.asInstanceOf[StringType]
-        if (st.supportsBinaryEquality) {
+        if (st.supportsBinaryEquality && !st.usesTrimCollation) {
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:569]: on another note, let's make sure that we reconsider this in the follow-up:
https://issues.apache.org/jira/browse/SPARK-49911

please add `// TODO(SPARK-49911)` comments wherever appropriate, just so that we don't miss it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 20]
```diff
@@ -817,7 +817,7 @@ case class HiveHash(children: Seq[Expression]) extends HashExpression[Int] {
 
   override protected def genHashString(
       ctx: CodegenContext, stringType: StringType, input: String, result: String): String = {
-    if (stringType.supportsBinaryEquality) {
+    if (stringType.supportsBinaryEquality && !stringType.usesTrimCollation) {
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:820]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala [hunk 21]
```diff
@@ -205,7 +205,9 @@ object UnsafeRowUtils {
    * can lead to rows being semantically equal even though their binary representations differ).
    */
   def isBinaryStable(dataType: DataType): Boolean = !dataType.existsRecursively {
-    case st: StringType => !CollationFactory.fetchCollation(st.collationId).supportsBinaryEquality
+    case st: StringType =>
+      val collation = CollationFactory.fetchCollation(st.collationId)
+      (!collation.supportsBinaryEquality || collation.supportsSpaceTrimming)
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:210]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala [hunk 22]
```diff
@@ -173,8 +173,9 @@ abstract class HashMapGenerator(
             ${hashBytes(bytes)}
           """
         }
-      case st: StringType if st.supportsBinaryEquality => hashBytes(s"$input.getBytes()")
-      case st: StringType if !st.supportsBinaryEquality =>
+      case st: StringType if st.supportsBinaryEquality && !st.usesTrimCollation =>
+        hashBytes(s"$input.getBytes()")
+      case st: StringType if !st.supportsBinaryEquality || st.usesTrimCollation =>
```
Comment (uros-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashMapGenerator.scala:178]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 18 above)
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: Yes we do, [here](https://github.com/apache/spark/pull/48386/files#diff-11ada8e3fe4dc4efd5e5ddf617e3ff405415f5a9341eb8f1054e79b4fa00538f:~:text=collation%22)%20%7B-,test(%22Support%20Md5%20hash%20expression%20with%20collation%22),-%7B) for example

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala (hunk 17 above)
Comment (uros-db) [reply sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:45]: maybe not at the moment, but I recommend that we keep it for private SQL datatype API completeness

let's not remove it for now, especially given that this change doesn't have anything to do with this PR

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:1]: here the number of tests is fine, since we don't want to explode the e2e sql tests

just one question - why did you add tests for Bin/Hex, but not for other expressions?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1]: what about hash joins?

shall we add some tests to:
test("hash join should be used for collated strings")

**Review by uros-db — CHANGES_REQUESTED** [review]
added a couple of comments, otherwise looks good

please address and request final re-review

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala
Comment (jovanpavl-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:1]: Yes, I added tests for hashings + accidentally for Bin and Hex.

Comment (jovanpavl-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:1]: Could remove it if you are hard opinionated about this.

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:1]: let's remove them for now, since they're not relevant here

we'll need to add tests for all expressions in this suite at one point anyway - so let's do Bin/Hex in that batch

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:1]: but a more important question is - why are these tests passing?

I was under the impression that these expressions don't yet support trimming sensitivity, and should hence fail in analysis due to unsupported collation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 19 above)
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:569]: I'm not really a fan of leaving comments like this in code. I will follow up with pr **today**, so won't miss them.

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala (hunk 16 above)
Comment (jovanpavl-db) [reply common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala]: added more tests.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 23]
```diff
@@ -917,16 +957,35 @@ protected Collation buildCollation() {
         Collator collator = Collator.getInstance(resultLocale);
         // Freeze ICU collator to ensure thread safety.
         collator.freeze();
+
+        Comparator<UTF8String> comparator;
+        ToLongFunction<UTF8String> hashFunction;
+
+        if (spaceTrimming == SpaceTrimming.NONE){
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:964]:
```suggestion
        if (spaceTrimming == SpaceTrimming.NONE) {
```

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala [hunk 24]
```diff
@@ -127,22 +127,38 @@ class CollationFactorySuite extends AnyFunSuite with Matchers { // scalastyle:ig
       CollationTestCase("UTF8_BINARY", "aaa", "AAA", false),
       CollationTestCase("UTF8_BINARY", "aaa", "bbb", false),
       CollationTestCase("UTF8_BINARY", "å", "a\u030A", false),
+      CollationTestCase("UTF8_BINARY_RTRIM", "aaa", "aaa", true),
```
Comment (uros-db) [inline common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:130]: I would like to see maybe a couple more tests here, like:
```
      CollationTestCase("UTF8_BINARY_RTRIM", "", "", true),
      CollationTestCase("UTF8_BINARY_RTRIM", "", "    ", true),
      CollationTestCase("UTF8_BINARY_RTRIM", "    ", "  ", true),
```
and also:
```
      CollationTestCase("UTF8_BINARY_RTRIM", "  aaa", "aaa", false),
```

**Review by uros-db — APPROVED** [review]
couple nits, otherwise lgtm

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @jovanpavl-db and @uros-db @mihailom-db for review.

**Comment by MaxGekk** [comment]
@jovanpavl-db Is https://issues.apache.org/jira/browse/SPARK-49661 correct JIRA for the PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 18 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: @jovanpavl-db Is the check `!stringType.usesTrimCollation` redundant, btw? `supportsBinaryEquality` should be false when trim is on, shouldn't it?

Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:418]: Correct, it happened because pr afterwards changed how supportsBinaryEquality is implemented. Will have cleanup pr for sure :)
